packages/indexer-common/src/indexer-management/allocations.ts

```diff
@@ -1148,4 +1157,73 @@ export class AllocationManager {
     return isDeploymentWorthAllocatingTowards(logger, subgraphDeployment, indexingRules)
       .toAllocate
   }
+
+  // Calculates the balance (GRT delta) of a single Action.
+  async resolveActionDelta(action: Action): Promise<ResolvedAction> {
+    let unallocates = BigNumber.from(0)
+
+    // Handle allocations
+    const allocates = BigNumber.from(action.amount ?? 0)
+
+    // Handle unallocations
+    if (action.type === ActionType.UNALLOCATE || action.type === ActionType.REALLOCATE) {
+      // ensure this Action have a valid allocationID
+      if (action.allocationID === null || action.allocationID === undefined) {
+        throw Error(
+          `SHOULD BE UNREACHABLE: Unallocate or Reallocate action must have an allocationID field: ${action}`,
+        )
+      }
+      // Fetch the allocation on chain to inspect its amount
+      const allocation = await this.networkMonitor.allocation(action.allocationID)
+      if (!allocation) {
+        throw indexerError(
+          IndexerErrorCode.IE063,
+          `Action validation failed: No active allocation found on chain with id '${action.allocationID}' found`,
+        )
+      }
+      /* We intentionally don't check if the allocation is active now because it will be checked
+       * later when we prepare the transaction */
+      unallocates = unallocates.add(allocation.allocatedTokens)
+    }
+
+    const balance = allocates.sub(unallocates)
+    return {
+      action,
+      allocates,
+      unallocates,
+      balance,
+    }
+  }
+
+  async validateActionBatchFeasibilty(batch: Action[]): Promise<Action[]> {
+    const logger = this.logger.child({ function: 'validateActionBatch' })
+    logger.debug(`Validating action batch`, { size: batch.length })
+
+    // Validate stake feasibility
+    const freeStake = await this.contracts.staking.getIndexerCapacity(this.indexer)
+    const mapper = async (action: Action) => this.resolveActionDelta(action)
+    const resolvedBatch = await pMap(batch, mapper)
+    const batchDelta: BigNumber = resolvedBatch
+      .map((resolvedAction) => resolvedAction.balance)
+      .reduce((a, b) => a.add(b))
+    const newBalance = batchDelta.add(freeStake)
+    if (newBalance.isNegative()) {
+      {
+        throw indexerError(
+          IndexerErrorCode.IE013, // TODO: Should we create a new error code?
```

Comment:
I could go either way on this.  Perhaps we try to use the existing error code, 13, and see if that is clear enough.